### PR TITLE
Re-land precompile triton templates

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124031
* #122825
* #123229
* #122642
* __->__ #124030

Re-land precompile triton templates. This got reverted because we were precompiling templates without checking the cache. I have since added logic and a test to ensure we do not precompile if there is a cache hit.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by shunting314 — APPROVED** [review]

**Review by nmacchioni — APPROVED** [review]

**Review by yoyoyocmu — APPROVED** [review]

**Comment by DanilBaibak** [comment]
Hi @eellison! Sorry, I need to revert your PR because it broke ROCm jobs. Here you can find more information about the issue - https://hud.pytorch.org/pytorch/pytorch/commit/d68196e7ef5eb8f62064ef70c75032f4d8b4a4fa

**Comment by DanilBaibak** [comment]
Hi @eellison! Sorry, I need to revert your PR because it broke trunk. Here you can find more information about the issue - https://hud.pytorch.org/pytorch/pytorch/commit/030bb13fe84c88ab5c988351543362b60fefb556

File: test/inductor/test_templated_attention.py [hunk 1]
```diff
@@ -16,7 +16,8 @@
 
 # Skip tests if Triton is not available
 supported_platform = skipUnless(
```
Comment (eellison) [inline test/inductor/test_templated_attention.py:18]: cc @jeffdaily

**Pull request closed without merging**